File: android/src/main/kotlin/com/bugreport/LogExporter.kt
package com.bugreport

import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Export logs in text, JSON, or CSV format — mirrors JS BR.exportLogs().
 */
class LogExporter {

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)

    fun export(logs: List<LogEntry>, stats: BugReportStats?, format: String = "json"): String {
        return when (format.lowercase()) {
            "text" -> exportText(logs, stats)
            "csv" -> exportCsv(logs)
            else -> exportJson(logs, stats)
        }
    }

    private fun exportText(logs: List<LogEntry>, stats: BugReportStats?): String {
        val sb = StringBuilder()
        sb.appendLine("BugReport Log Export (Android)")
        sb.appendLine("=".repeat(60))

        val device = logs.firstOrNull()?.device
        if (device != null) {
            sb.appendLine("Device:  ${device.brand} ${device.model}")
            sb.appendLine("System:  ${device.system} ${device.osVer}")
            sb.appendLine("App:     ${device.appName} v${device.appVer}")
        }
        sb.appendLine("Export:  ${dateFormat.format(Date())}")
        sb.appendLine("Logs:    ${logs.size}" +
            if (stats != null) " | Errors: ${stats.errors} | Warnings: ${stats.warnings}" else "")
        sb.appendLine("=".repeat(60))
        sb.appendLine()

        if (logs.isEmpty()) {
            sb.appendLine("No logs.")
            return sb.toString()
        }

        for (log in logs.reversed()) {
            val icon = when {
                log.level >= LogLevel.ERROR.value -> "[E]"
                log.level >= LogLevel.WARN.value -> "[W]"
                log.level >= LogLevel.INFO.value -> "[I]"
                else -> "[D]"
            }
            sb.appendLine("#${log.id} $icon ${log.time.substring(11, 23)} ${log.cat} ${log.tag}")
            sb.appendLine("  page: ${log.page}")
            if (log.msg.isNotEmpty()) sb.appendLine("  ${log.msg}")
            if (!log.stack.isNullOrEmpty()) {
                sb.append("  stack: ")
                log.stack.split("\n").take(3).forEach { sb.appendLine("  > $it") }
            }
            if (log.extra != null) {
                sb.appendLine("  data: ${JSONObject(log.extra)}")
            }
            sb.appendLine()
        }
        return sb.toString()
    }

    private fun exportJson(logs: List<LogEntry>, stats: BugReportStats?): String {
        val json = JSONObject()
        json.put("exportedAt", dateFormat.format(Date()))

        val device = logs.firstOrNull()?.device
        if (device != null) {
            json.put("device", JSONObject().apply {
                put("model", device.model)
                put("brand", device.brand)
                put("system", device.system)
                put("osVer", device.osVer)
                put("appVer", device.appVer)
                put("appName", device.appName)
            })
        }

        if (stats != null) {
            json.put("stats", JSONObject().apply {
                put("total", stats.total)
                put("errors", stats.errors)
                put("warnings", stats.warnings)
                put("fatals", stats.fatals)
                put("sessionMs", stats.sessionMs)
            })
        }

        val arr = JSONArray()
        for (log in logs) {
            arr.put(JSONObject().apply {
                put("id", log.id)
                put("ts", log.ts)
                put("time", log.time)
                put("level", log.level)
                put("levelLabel", log.levelLabel)
                put("cat", log.cat)
                put("tag", log.tag)
                put("msg", log.msg)
                put("stack", log.stack ?: "")
                put("page", log.page)
            })
        }
        json.put("logs", arr)
        return json.toString(2)
    }

    private fun exportCsv(logs: List<LogEntry>): String {
        val sb = StringBuilder()
        sb.appendLine("id,time,level,category,tag,message,page")
        for (log in logs) {
            sb.append("${log.id},${log.time},${log.levelLabel},${log.cat},${log.tag},")
            sb.append("\"${log.msg.replace("\"", "\"\"")}\",${log.page}")
            sb.appendLine()
        }
        return sb.toString()
    }
}

File: ios/Sources/BugReport/LogExporter.swift
import Foundation
import UIKit

/// Export logs in text, JSON, or CSV format — mirrors JS BR.exportLogs() and Kotlin LogExporter.
public class LogExporter {
    private let dateFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    public func export(_ logs: [LogEntry], stats: BugReportStats?, format: String = "json") -> String {
        switch format.lowercased() {
        case "text": return exportText(logs, stats: stats)
        case "csv":  return exportCsv(logs)
        default:     return exportJson(logs, stats: stats)
        }
    }

    private func exportText(_ logs: [LogEntry], stats: BugReportStats?) -> String {
        var sb = ""
        sb += "BugReport Log Export (iOS)\n"
        sb += String(repeating: "=", count: 60) + "\n"

        if let device = logs.first?.device {
            sb += "Device:  \(device.brand) \(device.model)\n"
            sb += "System:  \(device.system) \(device.osVer)\n"
            sb += "App:     \(device.appName) v\(device.appVer)\n"
        }
        sb += "Export:  \(dateFormatter.string(from: Date()))\n"
        sb += "Logs:    \(logs.count)"
        if let s = stats { sb += " | Errors: \(s.errors) | Warnings: \(s.warnings)" }
        sb += "\n" + String(repeating: "=", count: 60) + "\n\n"

        if logs.isEmpty { sb += "No logs.\n"; return sb }

        for log in logs.reversed() {
            let icon = log.level >= LogLevel.error.rawValue ? "[E]" :
                       log.level >= LogLevel.warn.rawValue ? "[W]" :
                       log.level >= LogLevel.info.rawValue ? "[I]" : "[D]"
            let t = String(log.time.suffix(12))
            sb += "#\(log.id) \(icon) \(t) \(log.cat) \(log.tag)\n"
            sb += "  page: \(log.page)\n"
            if !log.msg.isEmpty { sb += "  \(log.msg)\n" }
            if !log.stack.isEmpty {
                let lines = log.stack.split(separator: "\n").prefix(3)
                for line in lines { sb += "  > \(line)\n" }
            }
            sb += "\n"
        }
        return sb
    }

    private func exportJson(_ logs: [LogEntry], stats: BugReportStats?) -> String {
        var dict: [String: Any] = [
            "exportedAt": dateFormatter.string(from: Date()),
            "logs": logs.map { entry -> [String: Any] in
                [
                    "id": entry.id, "ts": entry.ts, "time": entry.time,
                    "level": entry.level, "levelLabel": entry.levelLabel,
                    "cat": entry.cat, "tag": entry.tag, "msg": entry.msg,
                    "stack": entry.stack, "page": entry.page
                ]
            }
        ]
        if let s = stats {
            dict["stats"] = ["total": s.total, "errors": s.errors, "warnings": s.warnings,
                             "fatals": s.fatals, "sessionMs": s.sessionMs]
        }
        guard let data = try? JSONSerialization.data(withJSONObject: dict, options: [.prettyPrinted]),
              let json = String(data: data, encoding: .utf8) else { return "{}" }
        return json
    }

    private func exportCsv(_ logs: [LogEntry]) -> String {
        var sb = "id,time,level,category,tag,message,page\n"
        for log in logs {
            let msg = log.msg.replacingOccurrences(of: "\"", with: "\"\"")
            sb += "\(log.id),\(log.time),\(log.levelLabel),\(log.cat),\(log.tag),\"\(msg)\",\(log.page)\n"
        }
        return sb
    }
}
